src/dev/lupio/LupioTTY.py

```diff
@@ -0,0 +1,55 @@
+# Copyright (c) 2021 The Regents of the University of California
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are
+# met: redistributions of source code must retain the above copyright
+# notice, this list of conditions and the following disclaimer;
+# redistributions in binary form must reproduce the above copyright
+# notice, this list of conditions and the following disclaimer in the
+# documentation and/or other materials provided with the distribution;
+# neither the name of the copyright holders nor the names of its
+# contributors may be used to endorse or promote products derived from
+# this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from m5.params import Param
+from m5.proxy import *
+from m5.util.fdthelper import *
+
+from m5.objects.Device import BasicPioDevice
+from m5.objects.Serial import SerialDevice
+
+class LupioTTY(BasicPioDevice):
+    type = 'LupioTTY'
+    cxx_class = 'gem5::LupioTTY'
+    cxx_header = "dev/lupio/lupio_tty.hh"
+    device = Param.SerialDevice(Parent.terminal_tty, "The terminal")
+#    device = Param.SerialDevice(Parent.any, "The terminal")
+    pio_size = Param.Addr(0x1000, "PIO size")
+    platform = Param.Platform(Parent.any,
+                              "Platform this device is part of.")
+
+    def generateDeviceTree(self, state):
+        node = self.generateBasicPioDeviceNode(state, "lupio-tty",
+                                               self.pio_addr,
+                                               self.pio_size)
+        platform = self.platform.unproxy(self)
+        plic = platform.plic
```

Comment:
Back to our convo from the other day, it does seem like this unproxying approach offers a certain level of genericity.

If the platform defined its interrupt controller with a more generic name (eg `intc` or `pic`), then this component (and same for the RNG) could have a line such as `pic = platform.pic`.

The only constraint for using the lupio devices from any platform would then to be that the platform defines a member called `pic`, but the interrupt controller itself could be of any kind (instead of implicitly restricting it to be a PLIC necessarily like the naming here strongly suggests).

In other words, until the new FDT code lands, I suggest to use this unproxying approach but to replace `plic` with `pic`.

Reply:
That's fine with me. Both approaches are awful in their own way. :)

Reply:
Should I still make this change now that we have the new FDT code?

Reply:
Yeah, the new code is going to take a few days to get in, and then you'll have to rebase. I vote that you shouldn't be blocked by that :)